presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java

```diff
@@ -469,4 +470,25 @@ public String toString()
                 .add("estimatedSize", estimatedSize)
                 .toString();
     }
+
+    public Iterator<Page> getPages(int[] outputChannels)
+    {
+        return new Iterator<Page>() {
+            int currentPosition = 0;
+            PageBuilder pageBuilder = new PageBuilder(types);
+            @Override
+            public boolean hasNext()
+            {
+                pageBuilder.reset();
+                currentPosition = buildPage(currentPosition, outputChannels, pageBuilder);
+                return !pageBuilder.isEmpty();
```

Comment:
This is not a correct implementation of the `Iterator`.

* `Iterator.hasNext()` must be idempotent
* `Iterator.hasNext()` doesn't need to be called. For the caller `!hasNext()`  only signals that `next()` would throw `NoSuchElementException`. Caller is free to use the iterator calling only `next()`.

Fear not though, there's still hope. Just extend `AbstractIterator` from Guava.